File: youtube-downloader/minio-service.js
import * as Minio from 'minio';
import fs from 'fs';
import path from 'path';
import { promisify } from 'util';

/**
 * Minio Service for file upload and download operations
 */
export class MinioService {
  constructor(config = {}) {
    // Default configuration
    const defaultConfig = {
      endPoint: 'localhost',
      port: 9000,
      useSSL: false,
      accessKey: 'minioadmin',
      secretKey: 'minioadmin'
    };

    this.config = { ...defaultConfig, ...config };
    this.client = new Minio.Client(this.config);
  }

  /**
   * Check if bucket exists, create if it doesn't
   * @param {string} bucketName - Name of the bucket
   * @returns {Promise<boolean>} - True if bucket exists or was created
   */
  async ensureBucket(bucketName) {
    try {
      const exists = await this.client.bucketExists(bucketName);
      if (!exists) {
        await this.client.makeBucket(bucketName);
        console.log(`🪣 Created bucket: ${bucketName}`);
      }
      return true;
    } catch (error) {
      console.error(`❌ Error ensuring bucket ${bucketName}:`, error.message);
      throw error;
    }
  }

  /**
   * Upload a file to Minio bucket
   * @param {string} bucketName - Name of the bucket
   * @param {string} objectName - Name of the object in the bucket (can include folders)
   * @param {string} filePath - Local file path to upload
   * @param {Object} metadata - Optional metadata for the object
   * @returns {Promise<Object>} - Upload result with etag and other info
   */
  async uploadFile(bucketName, objectName, filePath, metadata = {}) {
    try {
      // Ensure bucket exists
      await this.ensureBucket(bucketName);

      // Check if file exists locally
      if (!fs.existsSync(filePath)) {
        throw new Error(`File not found: ${filePath}`);
      }

      const fileStats = fs.statSync(filePath);
      console.log(`📤 Uploading ${filePath} (${(fileStats.size / 1024 / 1024).toFixed(2)} MB) to ${bucketName}/${objectName}`);

      // Add some default metadata
      const fullMetadata = {
        'Content-Type': this.getContentType(filePath),
        'upload-date': new Date().toISOString(),
        'file-size': fileStats.size.toString(),
        ...this.sanitizeMetadata(metadata)
      };

      const result = await this.client.fPutObject(bucketName, objectName, filePath, fullMetadata);
      
      console.log(`✅ Upload completed: ${bucketName}/${objectName}`);
      console.log(`📋 ETag: ${result.etag}`);

      return {
        success: true,
        etag: result.etag,
        bucketName,
        objectName,
        fileSize: fileStats.size,
        metadata: fullMetadata
      };

    } catch (error) {
      console.error(`❌ Upload failed for ${filePath}:`, error.message);
      throw error;
    }
  }

  /**
   * Download a file from Minio bucket
   * @param {string} bucketName - Name of the bucket
   * @param {string} objectName - Name of the object in the bucket
   * @param {string} downloadPath - Local path where file should be downloaded
   * @returns {Promise<Object>} - Download result
   */
  async downloadFile(bucketName, objectName, downloadPath) {
    try {
      // Ensure download directory exists
      const downloadDir = path.dirname(downloadPath);
      if (!fs.existsSync(downloadDir)) {
        fs.mkdirSync(downloadDir, { recursive: true });
      }

      console.log(`📥 Downloading ${bucketName}/${objectName} to ${downloadPath}`);

      await this.client.fGetObject(bucketName, objectName, downloadPath);

      const fileStats = fs.statSync(downloadPath);
      console.log(`✅ Download completed: ${downloadPath} (${(fileStats.size / 1024 / 1024).toFixed(2)} MB)`);

      return {
        success: true,
        bucketName,
        objectName,
        downloadPath,
        fileSize: fileStats.size
      };

    } catch (error) {
      console.error(`❌ Download failed for ${bucketName}/${objectName}:`, error.message);
      throw error;
    }
  }

  /**
   * List objects in a bucket with optional prefix (folder)
   * @param {string} bucketName - Name of the bucket
   * @param {string} prefix - Optional prefix to filter objects (folder path)
   * @param {boolean} recursive - Whether to list recursively
   * @returns {Promise<Array>} - Array of object information
   */
  async listObjects(bucketName, prefix = '', recursive = true) {
    try {
      const objects = [];
      const stream = this.client.listObjects(bucketName, prefix, recursive);

      return new Promise((resolve, reject) => {
        stream.on('data', (obj) => {
          objects.push({
            name: obj.name,
            size: obj.size,
            lastModified: obj.lastModified,
            etag: obj.etag
          });
        });

        stream.on('error', (error) => {
          console.error(`❌ Error listing objects in ${bucketName}:`, error.message);
          reject(error);
        });

        stream.on('end', () => {
          console.log(`📋 Found ${objects.length} objects in ${bucketName}${prefix ? `/${prefix}` : ''}`);
          resolve(objects);
        });
      });

    } catch (error) {
      console.error(`❌ Error listing objects:`, error.message);
      throw error;
    }
  }

  /**
   * Delete an object from bucket
   * @param {string} bucketName - Name of the bucket
   * @param {string} objectName - Name of the object to delete
   * @returns {Promise<boolean>} - True if deleted successfully
   */
  async deleteObject(bucketName, objectName) {
    try {
      await this.client.removeObject(bucketName, objectName);
      console.log(`🗑️  Deleted: ${bucketName}/${objectName}`);
      return true;
    } catch (error) {
      console.error(`❌ Delete failed for ${bucketName}/${objectName}:`, error.message);
      throw error;
    }
  }

  /**
   * Get object metadata
   * @param {string} bucketName - Name of the bucket
   * @param {string} objectName - Name of the object
   * @returns {Promise<Object>} - Object metadata
   */
  async getObjectMetadata(bucketName, objectName) {
    try {
      const stats = await this.client.statObject(bucketName, objectName);
      return {
        size: stats.size,
        lastModified: stats.lastModified,
        etag: stats.etag,
        metadata: stats.metaData
      };
    } catch (error) {
      // Don't log "not found" as errors - these are expected when files don't exist yet
      if (error.code === 'NotFound' || error.code === 'NoSuchKey' || error.message.includes('Not Found')) {
        throw error; // Re-throw without logging - this is expected behavior
      } else {
        console.error(`❌ Error getting metadata for ${bucketName}/${objectName}:`, error.message);
        throw error;
      }
    }
  }

  /**
   * Get a presigned URL for temporary access to an object
   * @param {string} bucketName - Name of the bucket
   * @param {string} objectName - Name of the object
   * @param {number} expiry - Expiry time in seconds (default: 1 hour)
   * @returns {Promise<string>} - Presigned URL
   */
  async getPresignedUrl(bucketName, objectName, expiry = 3600) {
    try {
      const url = await this.client.presignedGetObject(bucketName, objectName, expiry);
      console.log(`🔗 Generated presigned URL for ${bucketName}/${objectName} (expires in ${expiry}s)`);
      return url;
    } catch (error) {
      console.error(`❌ Error generating presigned URL:`, error.message);
      throw error;
    }
  }

  /**
   * Upload multiple files to a bucket/folder
   * @param {string} bucketName - Name of the bucket
   * @param {Array} files - Array of {filePath, objectName} objects
   * @param {string} folderPrefix - Optional folder prefix
   * @returns {Promise<Array>} - Array of upload results
   */
  async uploadMultipleFiles(bucketName, files, folderPrefix = '') {
    console.log(`🚀 Starting upload of ${files.length} files to ${bucketName}${folderPrefix ? `/${folderPrefix}` : ''}`);
    
    const results = [];
    for (let i = 0; i < files.length; i++) {
      const file = files[i];
      try {
        console.log(`\n📁 File ${i + 1}/${files.length}: ${file.filePath}`);
        
        const objectName = folderPrefix 
          ? `${folderPrefix}/${file.objectName || path.basename(file.filePath)}`
          : (file.objectName || path.basename(file.filePath));

        const result = await this.uploadFile(bucketName, objectName, file.filePath, file.metadata);
        results.push({ ...result, filePath: file.filePath });
      } catch (error) {
        console.error(`❌ Failed to upload ${file.filePath}:`, error.message);
        results.push({ 
          success: false, 
          filePath: file.filePath, 
          error: error.message 
        });
      }
    }

    // Summary
    const successful = results.filter(r => r.success).length;
    const failed = results.filter(r => !r.success).length;
    
    console.log(`\n📊 Upload Summary:`);
    console.log(`✅ Successful: ${successful}`);
    console.log(`❌ Failed: ${failed}`);

    return results;
  }

  /**
   * Determine content type based on file extension
   * @param {string} filePath - File path
   * @returns {string} - Content type
   */
  getContentType(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    const contentTypes = {
      '.mp4': 'video/mp4',
      '.mp3': 'audio/mpeg',
      '.wav': 'audio/wav',
      '.avi': 'video/x-msvideo',
      '.mov': 'video/quicktime',
      '.mkv': 'video/x-matroska',
      '.json': 'application/json',
      '.txt': 'text/plain',
      '.pdf': 'application/pdf',
      '.jpg': 'image/jpeg',
      '.jpeg': 'image/jpeg',
      '.png': 'image/png',
      '.gif': 'image/gif'
    };

    return contentTypes[ext] || 'application/octet-stream';
  }

  /**
   * Test connection to Minio server
   * @returns {Promise<boolean>} - True if connection is successful
   */
  async testConnection() {
    try {
      // Try to list buckets as a connection test
      await this.client.listBuckets();
      console.log('✅ Minio connection successful');
      return true;
    } catch (error) {
      console.error('❌ Minio connection failed:', error.message);
      return false;
    }
  }

  /**
   * Sanitize metadata values for HTTP headers
   * @param {Object} metadata - Raw metadata object
   * @returns {Object} - Sanitized metadata safe for HTTP headers
   */
  sanitizeMetadata(metadata) {
    const sanitized = {};
    
    for (const [key, value] of Object.entries(metadata)) {
      if (value != null) {
        // Sanitize the value to be safe for HTTP headers
        sanitized[key] = String(value)
          .replace(/[^\x20-\x7E]/g, '')     // Remove non-ASCII printable characters
          .replace(/[\r\n\t]/g, ' ')        // Replace line breaks and tabs with spaces
          .replace(/\s+/g, ' ')             // Collapse multiple spaces
          .trim()
          .substring(0, 255);               // Limit length for header safety
      }
    }
    
    return sanitized;
  }
}

export default MinioService;

File: services/minio_service.py
import os
import io
from datetime import datetime
from typing import Optional, Dict, Any
from minio import Minio
from minio.error import S3Error


class MinIOService:
    """
    A robust MinIO service for saving and retrieving data with folder organization.
    """

    def __init__(
        self,
        endpoint: str,
        access_key: str,
        secret_key: str,
        bucket_name: str,
        secure: bool = True,
        region: Optional[str] = None,
    ):
        """
        Initialize MinIO service with connection parameters.

        Args:
            endpoint: MinIO server endpoint (e.g., 'localhost:9000')
            access_key: Access key for authentication
            secret_key: Secret key for authentication
            bucket_name: Default bucket name to use
            secure: Whether to use HTTPS (default: True)
            region: Optional region name
        """
        self.endpoint = endpoint
        self.access_key = access_key
        self.secret_key = secret_key
        self.bucket_name = bucket_name
        self.secure = secure
        self.region = region

        # Initialize MinIO client
        self.client = Minio(
            endpoint=self.endpoint,
            access_key=self.access_key,
            secret_key=self.secret_key,
            secure=self.secure,
            region=self.region,
        )

        # Ensure bucket exists
        self._ensure_bucket_exists()

    def _ensure_bucket_exists(self) -> None:
        """
        Ensure the bucket exists, create if it doesn't.
        """
        try:
            if not self.client.bucket_exists(self.bucket_name):
                self.client.make_bucket(self.bucket_name, location=self.region)
                print(f"Created bucket: {self.bucket_name}")
        except S3Error as e:
            raise Exception(f"Failed to create/verify bucket {self.bucket_name}: {e}")

    def save(
        self,
        data: bytes,
        filename: str,
        content_type: str,
        bucket_name: Optional[str] = None,
        metadata: Optional[Dict[str, str]] = None,
    ) -> bool:
        """
        Save data to MinIO with folder organization.

        Args:
            data: Binary data to save
            filename: File name (e.g., 'AAPL_1min.csv')
            bucket_name: Optional bucket name override
            metadata: Optional metadata dictionary

        Returns:
            bool: True if successful, False otherwise
        """
        try:
            bucket = bucket_name or self.bucket_name

            # Convert bytes to BytesIO stream
            data_stream = io.BytesIO(data)

            # Upload the data
            self.client.put_object(
                bucket_name=bucket,
                object_name=filename,
                data=data_stream,
                length=len(data),
                content_type=content_type,
                metadata=metadata,
            )

            print(f"Successfully saved {filename} to bucket {bucket}")
            return True

        except S3Error as e:
            print(f"Failed to save {filename}: {e}")
            return False
        except Exception as e:
            print(f"Unexpected error saving {filename}: {e}")
            return False

    def save_file(
        self,
        file_path: str,
        folder: str,
        filename: Optional[str] = None,
        bucket_name: Optional[str] = None,
        metadata: Optional[Dict[str, str]] = None,
    ) -> bool:
        """
        Save a local file to MinIO.

        Args:
            file_path: Path to local file
            folder: Destination folder path
            filename: Optional custom filename (uses original if None)
            bucket_name: Optional bucket name override
            metadata: Optional metadata dictionary

        Returns:
            bool: True if successful, False otherwise
        """
        try:
            if not os.path.exists(file_path):
                print(f"File not found: {file_path}")
                return False

            # Use original filename if not provided
            if filename is None:
                filename = os.path.basename(file_path)

            bucket = bucket_name or self.bucket_name
            object_name = f"{folder.strip('/')}/{filename}"

            # Upload the file
            self.client.fput_object(
                bucket_name=bucket,
                object_name=object_name,
                file_path=file_path,
                metadata=metadata,
            )

            print(f"Successfully uploaded {file_path} as {object_name}")
            return True

        except S3Error as e:
            print(f"Failed to upload {file_path}: {e}")
            return False
        except Exception as e:
            print(f"Unexpected error uploading {file_path}: {e}")
            return False

    def retrieve(
        self, folder: str, filename: str, bucket_name: Optional[str] = None
    ) -> Optional[bytes]:
        """
        Retrieve data from MinIO.

        Args:
            folder: Folder path
            filename: File name
            bucket_name: Optional bucket name override

        Returns:
            bytes: File data if successful, None otherwise
        """
        try:
            bucket = bucket_name or self.bucket_name
            object_name = f"{folder.strip('/')}/{filename}"

            response = self.client.get_object(bucket, object_name)
            data = response.read()
            response.close()
            response.release_conn()

            return data

        except S3Error as e:
            print(f"Failed to retrieve {folder}/{filename}: {e}")
            return None
        except Exception as e:
            print(f"Unexpected error retrieving {folder}/{filename}: {e}")
            return None

    def retrieve_to_file(
        self,
        folder: str,
        filename: str,
        local_path: str,
        bucket_name: Optional[str] = None,
    ) -> bool:
        """
        Retrieve data from MinIO and save to local file.

        Args:
            folder: Folder path in MinIO
            filename: File name in MinIO
            local_path: Local file path to save to
            bucket_name: Optional bucket name override

        Returns:
            bool: True if successful, False otherwise
        """
        try:
            bucket = bucket_name or self.bucket_name
            object_name = f"{folder.strip('/')}/{filename}"

            # Ensure local directory exists
            os.makedirs(os.path.dirname(local_path), exist_ok=True)

            self.client.fget_object(bucket, object_name, local_path)

            print(f"Successfully downloaded {object_name} to {local_path}")
            return True

        except S3Error as e:
            print(f"Failed to download {folder}/{filename}: {e}")
            return False
        except Exception as e:
            print(f"Unexpected error downloading {folder}/{filename}: {e}")
            return False

    def save_data_with_date(
        self,
        data: bytes,
        date: datetime,
        category: str,
        filename: str,
        bucket_name: Optional[str] = None,
        metadata: Optional[Dict[str, str]] = None,
    ) -> bool:
        """
        Save data organized by date and category.

        Args:
            data: Binary data to save
            date: Date for folder organization
            category: Category subfolder (e.g., 'stocks', 'crypto')
            filename: File name
            bucket_name: Optional bucket name override
            metadata: Optional metadata dictionary

        Returns:
            bool: True if successful, False otherwise
        """
        folder = f"{category}/{date.strftime('%Y-%m-%d')}"
        return self.save(data, folder, filename, bucket_name, metadata)

    def retrieve_data_with_date(
        self,
        date: datetime,
        category: str,
        filename: str,
        bucket_name: Optional[str] = None,
    ) -> Optional[bytes]:
        """
        Retrieve data organized by date and category.

        Args:
            date: Date for folder lookup
            category: Category subfolder
            filename: File name
            bucket_name: Optional bucket name override

        Returns:
            bytes: File data if successful, None otherwise
        """
        folder = f"{category}/{date.strftime('%Y-%m-%d')}"
        return self.retrieve(folder, filename, bucket_name)

    def list_objects(
        self,
        folder: str = "",
        bucket_name: Optional[str] = None,
        recursive: bool = False,
    ) -> list:
        """
        List objects in a folder.

        Args:
            folder: Folder path (empty for root)
            bucket_name: Optional bucket name override
            recursive: Whether to list recursively

        Returns:
            list: List of object names
        """
        try:
            bucket = bucket_name or self.bucket_name
            prefix = folder.strip("/") + "/" if folder else ""

            objects = self.client.list_objects(
                bucket, prefix=prefix, recursive=recursive
            )

            return [obj.object_name for obj in objects]

        except S3Error as e:
            print(f"Failed to list objects in {folder}: {e}")
            return []
        except Exception as e:
            print(f"Unexpected error listing objects in {folder}: {e}")
            return []

    def delete_object(
        self, folder: str, filename: str, bucket_name: Optional[str] = None
    ) -> bool:
        """
        Delete an object from MinIO.

        Args:
            folder: Folder path
            filename: File name
            bucket_name: Optional bucket name override

        Returns:
            bool: True if successful, False otherwise
        """
        try:
            bucket = bucket_name or self.bucket_name
            object_name = f"{folder.strip('/')}/{filename}"

            self.client.remove_object(bucket, object_name)

            print(f"Successfully deleted {object_name}")
            return True

        except S3Error as e:
            print(f"Failed to delete {folder}/{filename}: {e}")
            return False
        except Exception as e:
            print(f"Unexpected error deleting {folder}/{filename}: {e}")
            return False

    def object_exists(
        self, folder: str, filename: str, bucket_name: Optional[str] = None
    ) -> bool:
        """
        Check if an object exists in MinIO.

        Args:
            folder: Folder path
            filename: File name
            bucket_name: Optional bucket name override

        Returns:
            bool: True if object exists, False otherwise
        """
        try:
            bucket = bucket_name or self.bucket_name
            object_name = f"{folder.strip('/')}/{filename}"

            self.client.stat_object(bucket, object_name)
            return True

        except S3Error:
            return False
        except Exception:
            return False

    def get_object_info(
        self, folder: str, filename: str, bucket_name: Optional[str] = None
    ) -> Optional[Dict[str, Any]]:
        """
        Get object information including metadata.

        Args:
            folder: Folder path
            filename: File name
            bucket_name: Optional bucket name override

        Returns:
            dict: Object info if successful, None otherwise
        """
        try:
            bucket = bucket_name or self.bucket_name
            object_name = f"{folder.strip('/')}/{filename}"

            stat = self.client.stat_object(bucket, object_name)

            return {
                "object_name": stat.object_name,
                "size": stat.size,
                "etag": stat.etag,
                "last_modified": stat.last_modified,
                "metadata": stat.metadata,
                "content_type": stat.content_type,
            }

        except S3Error as e:
            print(f"Failed to get info for {folder}/{filename}: {e}")
            return None
        except Exception as e:
            print(f"Unexpected error getting info for {folder}/{filename}: {e}")
            return None
